File: los-platform/backend/notification-service/src/controllers/notification.controller.ts
import {
  Controller,
  Post,
  Get,
  Put,
  Delete,
  Body,
  Param,
  Query,
  Headers,
  UseGuards,
  HttpCode,
  HttpStatus,
  ValidationPipe,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiBearerAuth, ApiHeader, ApiQuery } from '@nestjs/swagger';
import { JwtAuthGuard } from '@los/common';
import { NotificationService } from '../services/notification.service';
import {
  SendNotificationDto,
  BulkNotificationDto,
  BulkNotificationResponseDto,
  NotificationResponseDto,
  NotificationQueryDto,
  CreateTemplateDto,
  UpdateTemplateDto,
  UpdatePreferenceDto,
  OptOutDto,
  WebhookCallbackDto,
  OTPSendDto,
  OTPVerifyDto,
  NotificationStatsDto,
} from '../dto/notification.dto';
import { NotificationChannel, NotificationCategory } from '../entities/notification.entity';

@ApiTags('Notifications')
@Controller('notifications')
export class NotificationController {
  constructor(private readonly notificationService: NotificationService) {}

  @Post('send')
  @HttpCode(HttpStatus.OK)
  @ApiBearerAuth()
  @UseGuards(JwtAuthGuard)
  @ApiOperation({ summary: 'Send a single notification' })
  @ApiResponse({ status: 200, description: 'Notification sent', type: NotificationResponseDto })
  async send(@Body(ValidationPipe) dto: SendNotificationDto) {
    return this.notificationService.send(dto);
  }

  @Post('send/bulk')
  @HttpCode(HttpStatus.OK)
  @ApiBearerAuth()
  @UseGuards(JwtAuthGuard)
  @ApiOperation({ summary: 'Send bulk notifications' })
  @ApiResponse({ status: 200, description: 'Bulk notification results', type: BulkNotificationResponseDto })
  async sendBulk(@Body(ValidationPipe) dto: BulkNotificationDto) {
    return this.notificationService.sendBulk(dto);
  }

  @Post('otp/send')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Send OTP via SMS', description: 'OTP for login/consent/eSign/bureau_consent' })
  @ApiResponse({ status: 200, description: 'OTP sent successfully' })
  async sendOTP(@Body(ValidationPipe) dto: OTPSendDto) {
    const result = await this.notificationService.sendOTP(dto);
    return {
      success: result.success,
      messageId: result.messageId,
      expiresAt: result.expiresAt,
      maskedMobile: dto.mobileNumber.substring(0, 3) + 'XXXXX' + dto.mobileNumber.slice(-2),
    };
  }

  @Post('otp/verify')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Verify OTP' })
  @ApiResponse({ status: 200, description: 'OTP verification result' })
  @ApiResponse({ status: 400, description: 'Invalid or expired OTP' })
  async verifyOTP(@Body(ValidationPipe) dto: OTPVerifyDto) {
    const result = await this.notificationService.verifyOTP(dto);
    return result;
  }

  @Get('history')
  @ApiBearerAuth()
  @UseGuards(JwtAuthGuard)
  @ApiOperation({ summary: 'Get notification history with pagination' })
  @ApiResponse({ status: 200, description: 'Paginated notification history' })
  async getHistory(@Query(ValidationPipe) query: NotificationQueryDto) {
    const { notifications, total } = await this.notificationService.getNotificationHistory(query);
    const page = query.page || 1;
    const limit = query.limit || 20;
    return {
      notifications: notifications.map(n => ({
        id: n.id,
        applicationId: n.applicationId,
        channel: n.channel,
        category: n.category,
        status: n.status,
        recipientId: n.recipientId,
        templateName: n.templateName,
        subject: n.subject,
        renderedContent: n.renderedContent?.substring(0, 200),
        sentAt: n.sentAt,
        deliveredAt: n.deliveredAt,
        readAt: n.readAt,
        errorCode: n.errorCode,
        createdAt: n.createdAt,
      })),
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  @Post('webhook')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Handle provider delivery status webhook (no auth for external providers)' })
  async handleWebhook(@Body(ValidationPipe) dto: WebhookCallbackDto) {
    await this.notificationService.handleWebhook(dto);
    return { received: true };
  }

  @Post('opt-out')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Opt out from notifications' })
  async optOut(
    @Body(ValidationPipe) dto: OptOutDto,
    @Headers('x-forwarded-for') ip?: string,
  ) {
    await this.notificationService.optOut(dto.recipientId || '', dto, ip);
    return { success: true, message: `Successfully opted out from ${dto.channel} notifications` };
  }

  @Get('stats')
  @ApiBearerAuth()
  @UseGuards(JwtAuthGuard)
  @ApiOperation({ summary: 'Get notification delivery statistics' })
  @ApiResponse({ status: 200, description: 'Notification delivery statistics', type: NotificationStatsDto })
  @ApiQuery({ name: 'fromDate', required: false, description: 'Start date (ISO format)' })
  @ApiQuery({ name: 'toDate', required: false, description: 'End date (ISO format)' })
  async getStats(@Query('fromDate') fromDate?: string, @Query('toDate') toDate?: string): Promise<NotificationStatsDto> {
    return this.notificationService.getStats(fromDate, toDate);
  }
}

@ApiTags('Templates')
@ApiBearerAuth()
@UseGuards(JwtAuthGuard)
@Controller('notifications/templates')
export class TemplateController {
  constructor(private readonly notificationService: NotificationService) {}

  @Post()
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: 'Create a new notification template' })
  @ApiResponse({ status: 201, description: 'Template created' })
  async createTemplate(@Body(ValidationPipe) dto: CreateTemplateDto) {
    return this.notificationService.createTemplate(dto);
  }

  @Get()
  @ApiOperation({ summary: 'List all notification templates' })
  @ApiResponse({ status: 200, description: 'List of notification templates' })
  @ApiQuery({ name: 'channel', required: false, enum: NotificationChannel })
  @ApiQuery({ name: 'category', required: false, enum: NotificationCategory })
  async getTemplates(
    @Query('channel') channel?: NotificationChannel,
    @Query('category') category?: NotificationCategory,
  ) {
    return this.notificationService.getTemplates(channel, category);
  }

  @Put(':templateName')
  @ApiOperation({ summary: 'Update a notification template' })
  async updateTemplate(
    @Param('templateName') templateName: string,
    @Body(ValidationPipe) dto: UpdateTemplateDto,
  ) {
    return this.notificationService.updateTemplate(templateName, dto);
  }

  @Delete(':templateName')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Deactivate a notification template' })
  async deactivateTemplate(@Param('templateName') templateName: string) {
    await this.notificationService.updateTemplate(templateName, { isActive: false } as UpdateTemplateDto);
  }
}

@ApiTags('Preferences')
@ApiBearerAuth()
@UseGuards(JwtAuthGuard)
@Controller('notifications/preferences')
export class PreferenceController {
  constructor(private readonly notificationService: NotificationService) {}

  @Put()
  @ApiOperation({ summary: 'Update notification preferences for a user' })
  @ApiResponse({ status: 200, description: 'Preference updated' })
  async updatePreference(
    @Headers('x-user-id') userId: string,
    @Body(ValidationPipe) dto: UpdatePreferenceDto,
  ) {
    return this.notificationService.updatePreference(userId, dto);
  }
}

File: los-platform/backend-java/src/main/java/com/los/notification/controller/NotificationController.java
package com.los.notification.controller;

import com.los.notification.dto.*;
import com.los.notification.service.NotificationService;
import com.los.common.dto.ApiResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Slf4j
@RestController
@RequestMapping("/api/notifications")
@RequiredArgsConstructor
@Tag(name = "Notification", description = "Notification APIs for SMS, Email, WhatsApp, Push")
@SecurityRequirement(name = "bearerAuth")
public class NotificationController {

    private final NotificationService notificationService;

    @PostMapping("/send")
    @Operation(summary = "Send notification", description = "Send notification via SMS/Email/WhatsApp/Push")
    public ResponseEntity<ApiResponse<NotificationStatusDto>> sendNotification(
            @Valid @RequestBody SendNotificationDto dto) {
        log.info("POST /api/notifications/send - Recipient: {}, Channel: {}", dto.getRecipientId(), dto.getChannel());

        NotificationStatusDto response = notificationService.sendNotification(dto);
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(response, "Notification sent successfully"));
    }

    @GetMapping("/status/{id}")
    @Operation(summary = "Check notification status", description = "Get status of sent notification")
    public ResponseEntity<ApiResponse<NotificationStatusDto>> getNotificationStatus(
            @PathVariable String id) {
        log.info("GET /api/notifications/status/{} - Fetching status", id);

        NotificationStatusDto response = notificationService.getNotificationStatus(id);
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(response, "Notification status retrieved"));
    }

    @PostMapping("/template")
    @Operation(summary = "Create notification template", description = "Create template for admin")
    public ResponseEntity<ApiResponse<TemplateDataDto>> createTemplate(
            @Valid @RequestBody TemplateDataDto dto) {
        log.info("POST /api/notifications/template - Creating template: {}", dto.getTemplateName());

        TemplateDataDto response = notificationService.createTemplate(dto);
        return ResponseEntity
                .status(HttpStatus.CREATED)
                .body(ApiResponse.success(response, "Template created successfully"));
    }

    @GetMapping("/templates")
    @Operation(summary = "List notification templates", description = "Get all active templates")
    public ResponseEntity<ApiResponse<List<TemplateDataDto>>> listTemplates() {
        log.info("GET /api/notifications/templates - Listing templates");

        List<TemplateDataDto> response = notificationService.listTemplates();
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(response, "Templates retrieved successfully"));
    }

    @GetMapping("/history/{recipientId}")
    @Operation(summary = "Get notification history", description = "Get all notifications for recipient")
    public ResponseEntity<ApiResponse<List<NotificationStatusDto>>> getNotificationHistory(
            @PathVariable String recipientId) {
        log.info("GET /api/notifications/history/{} - Fetching history", recipientId);

        List<NotificationStatusDto> response = notificationService.getNotificationHistory(recipientId);
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(response, "Notification history retrieved"));
    }
}
